include/unifex/task.hpp

```diff
@@ -195,23 +209,31 @@ struct _awaiter {
 
     coro::coroutine_handle<ThisPromise> await_suspend(
         coro::coroutine_handle<OtherPromise> h) noexcept {
-      UNIFEX_ASSERT(coro_);
-      auto& promise = coro_.promise();
+      UNIFEX_ASSERT(coro_ && ((coro_ & 1u) == 0u));
+      auto thisCoro = coro::coroutine_handle<ThisPromise>::from_address((void*) coro_);
+      ++coro_; // mark the awaiter as needing cleanup
+      auto& promise = thisCoro.promise();
+      sched_.construct(get_scheduler(h.promise()));
       promise.continuation_ = h;
       promise.stoken_ = stopTokenAdapter_.subscribe(get_stop_token(h.promise()));
-      return coro_;
+      promise.sched_ = sched_.get();
+      return thisCoro;
     }
 
     result_type await_resume() {
       stopTokenAdapter_.unsubscribe();
-      scope_guard destroyOnExit{[this]() noexcept { std::exchange(coro_, {}).destroy(); }};
-      return coro_.promise().result();
+      sched_.destruct();
+      auto thisCoro = coro::coroutine_handle<ThisPromise>::from_address(
+          (void*) std::exchange(--coro_, 0));
+      scope_guard destroyOnExit{[&]() noexcept { thisCoro.destroy(); }};
+      return thisCoro.promise().result();
     }
 
   private:
-    coro::coroutine_handle<ThisPromise> coro_;
+    std::uintptr_t coro_; // Stored as an integer so we can use the low bit as a dirty bit
```

Comment:
> If you get into the falsey | 1 state, I think you'll try to `.destroy()` an invalid `coroutine_handle<>`.

I don't think it's possible to get into this state.

> are trying to make it defined behaviour to construct and destruct an awaiter without awaiting it (why?)

Because it is possible to call `get_awaiter(task)` and ignore the result. Unlikely, but possible.

> need to destroy the `coroutine_handle<>` if it's truthy even if the dirty bit is false, and destroy the other state when the dirty bit is true.

Good catch.